diff --git a/lib/galaxy/workflow/run.py b/lib/galaxy/workflow/run.py
index 8025c10..e71ffd3 100644
--- a/lib/galaxy/workflow/run.py
+++ b/lib/galaxy/workflow/run.py
@@ -162,7 +162,7 @@ class WorkflowInvoker( object ):
             except modules.DelayedWorkflowEvaluation:
                 delayed_steps = True
                 self.progress.mark_step_outputs_delayed( step )
-            log.debug("Workflow step %s invoked %s" % (step.id, step_timer))
+            #log.debug("Workflow step %s invoked %s" % (step.id, step_timer))
 
         if delayed_steps:
             state = model.WorkflowInvocation.states.READY
