diff --git a/lib/galaxy/tools/execute.py b/lib/galaxy/tools/execute.py
index e2ba87b..76f6a94 100644
--- a/lib/galaxy/tools/execute.py
+++ b/lib/galaxy/tools/execute.py
@@ -72,6 +72,11 @@ def execute( trans, tool, param_combinations, history, rerun_remap_job_id=None,
     log.debug("Executed %d job(s) for tool %s request: %s" % (job_count, tool.id, all_jobs_timer))
     if collection_info:
         history = history or tool.get_default_history_by_trans( trans )
+        if len(param_combinations) == 0:
+            template = "Attempting to map over an empty collection, this is not yet implemented. colleciton_info is [%s]"
+            message = template % collection_info
+            log.warn(message)
+            raise Exception(message)
         params = param_combinations[0]
         execution_tracker.create_output_collections( trans, history, params )
 
