diff --git a/lib/galaxy/web/framework/decorators.py b/lib/galaxy/web/framework/decorators.py
index 4460b49..5919dab 100644
--- a/lib/galaxy/web/framework/decorators.py
+++ b/lib/galaxy/web/framework/decorators.py
@@ -149,7 +149,10 @@ def expose_api( func, to_json=True, user_required=True ):
 
 
 def __extract_payload_from_request(trans, func, kwargs):
-    content_type = trans.request.headers['content-type']
+    try:
+        content_type = trans.request.headers['content-type']
+    except KeyError:
+        content_type = 'application/json'
     if content_type.startswith('application/x-www-form-urlencoded') or content_type.startswith('multipart/form-data'):
         # If the content type is a standard type such as multipart/form-data, the wsgi framework parses the request body
         # and loads all field values into kwargs. However, kwargs also contains formal method parameters etc. which
